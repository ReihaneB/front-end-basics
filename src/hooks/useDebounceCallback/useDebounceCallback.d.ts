export type useDebounceCallbackProps = {
  callback: (...rest: unknown[]) => unknown;
  waitTime?: number;
  options?: {
    leading?: boolean;
    trailing?: boolean;
  };
  disableDebounce?: boolean;
};
