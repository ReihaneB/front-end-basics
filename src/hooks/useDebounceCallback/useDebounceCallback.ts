import { useCallback, useRef, useEffect } from 'react';

import { useDebounceCallbackProps } from './useDebounceCallback.d';

export function useDebounceCallback({
  callback,
  waitTime = 500,
  options = {
    leading: false,
    trailing: true,
  },
  disableDebounce = false,
}: useDebounceCallbackProps) {
  const { leading, trailing } = options;

  const callbackRef = useRef<(...rest: never[]) => void>(callback);
  callbackRef.current = callback;

  const timeoutRef = useRef<ReturnType<typeof setTimeout> | null>(null);
  const isLeadingCallRef = useRef<boolean>(false);

  const debounced = useCallback((...rest: never[]) => {
    if (disableDebounce) {
      callbackRef.current(...rest);
      return;
    }

    const isLeading = leading && !isLeadingCallRef.current;
    if (isLeading) {
      callbackRef.current(...rest);
      isLeadingCallRef.current = true;
    }

    if (trailing) {
      if (timeoutRef.current) {
        clearTimeout(timeoutRef.current);
      }

      timeoutRef.current = setTimeout(() => {
        isLeadingCallRef.current = false;
        callbackRef.current(...rest);
      }, waitTime);
    }
  }, [
    waitTime,
    leading,
    trailing,
    disableDebounce,
  ]);

  useEffect(() => function clearDebounce() {
    if (timeoutRef.current) {
      clearTimeout(timeoutRef.current);
    }
  }, []);

  return debounced;
}
