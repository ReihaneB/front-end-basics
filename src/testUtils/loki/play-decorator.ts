/* eslint-disable @typescript-eslint/ban-ts-comment */
// @ts-expect-error
import createAsyncCallback from '@loki/create-async-callback';
// @ts-expect-error
import isLokiRunning from '@loki/is-loki-running';
import type {PlayFunction} from '@storybook/csf';
// @ts-expect-error
import type {ReactFramework} from '@storybook/react';

async function waitForDocumentLoaded(): Promise<void> {
  if (document.readyState === 'loading') {
    return new Promise(resolve => {
      document.addEventListener('DOMContentLoaded', () => resolve());
    });
  }

  return Promise.resolve();
}

function enablePointerEvents(): void {
  const styleElement = document.createElement('style');
  document.documentElement.appendChild(styleElement);
  styleElement.sheet?.insertRule('* {pointer-events: auto !important}');
}

export function lokiPlayDecorator(
  target: PlayFunction<ReactFramework>
): PlayFunction<ReactFramework> {
  return async context => {
    const lokiRunning = isLokiRunning();
    const asyncCallback = createAsyncCallback();

    if (lokiRunning) {
      await waitForDocumentLoaded();
      enablePointerEvents();
    }

    try {
      await target(context);
    } finally {
      asyncCallback();
    }
  };
}
