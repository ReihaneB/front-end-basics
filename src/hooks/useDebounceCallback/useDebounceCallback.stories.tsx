import { useState } from 'react';
import { expect, userEvent, waitFor, within } from '@storybook/test';

import { lokiPlayDecorator } from '@/testUtils/loki/play-decorator';

import { useDebounceCallback } from './useDebounceCallback';

export default {
  title: 'Hooks/useDebounceCallback',
  component: useDebounceCallback,
};

function ClickableComponent({
  useDebounceCallbackProps = {},
}) {
  const [count, setCount] = useState(0);

  const handleClick = () => {
    setCount(prevState => prevState + 1);
  };

  const debounced = useDebounceCallback({ callback: handleClick, ...useDebounceCallbackProps });

  return (
    <div>
      <button
        type="button"
        data-testid="clickable"
        onClick={debounced}
      >
        Click me
      </button>
      <p
        data-testid="count"
      >
        Callback called
        {' '}
        {count}
        {' '}
        times
      </p>
    </div>
  );
}

export function Playground({...args}) {
  return <ClickableComponent {...args} />;
}

export function ClickMulipleTimes() {
  return <ClickableComponent />;
}

ClickMulipleTimes.storyName = 'Calls the callback only once after multiple calls';
ClickMulipleTimes.play = lokiPlayDecorator(
  async ({ canvasElement }: { canvasElement: HTMLElement }) => {
    const canvas = within(canvasElement);

    await userEvent.dblClick(canvas.getByTestId('clickable'));
    await userEvent.dblClick(canvas.getByTestId('clickable'));
    await userEvent.dblClick(canvas.getByTestId('clickable'));


    await waitFor(() => expect(canvas.getByTestId('count')).toHaveTextContent('Callback called 1 times'));
  }
);

export function CallAfterWaitTime() {
  return <ClickableComponent />;
}

CallAfterWaitTime.storyName = 'Calls the callback after the wait time';
CallAfterWaitTime.play = lokiPlayDecorator(
  async ({ canvasElement }: { canvasElement: HTMLElement }) => {
    const canvas = within(canvasElement);

    await userEvent.click(canvas.getByTestId('clickable'));
    await userEvent.click(canvas.getByTestId('clickable'));
    await userEvent.click(canvas.getByTestId('clickable'));
    await userEvent.click(canvas.getByTestId('clickable'));

    await waitFor(() => expect(canvas.getByTestId('count')).toHaveTextContent('Callback called 1 times'));
  }
);

export function DoNotCallBeforeWaitTime() {
  return <ClickableComponent />;
}

DoNotCallBeforeWaitTime.storyName = 'Does not call the callback before the wait time';
DoNotCallBeforeWaitTime.play = lokiPlayDecorator(
  async ({ canvasElement }: { canvasElement: HTMLElement }) => {
    const canvas = within(canvasElement);

    await userEvent.click(canvas.getByTestId('clickable'));

    await waitFor(() => expect(canvas.getByTestId('count')).toHaveTextContent('Callback called 0 times'), { timeout: 100 });
    await waitFor(() => expect(canvas.getByTestId('count')).toHaveTextContent('Callback called 1 times'));
  }
);

export function LeadingTrue() {
  return (
    <ClickableComponent useDebounceCallbackProps={{ options: { leading: true } }} />
  );
}

LeadingTrue.storyName = 'Calls the callback immediately if leading is true';
LeadingTrue.play = lokiPlayDecorator(
  async ({ canvasElement }: { canvasElement: HTMLElement }) => {
    const canvas = within(canvasElement);

    await userEvent.click(canvas.getByTestId('clickable'));

    await waitFor(() => expect(canvas.getByTestId('count')).toHaveTextContent('Callback called 1 times'), { timeout: 0 });
  }
);

export function TrailingTrue() {
  return (
    <ClickableComponent useDebounceCallbackProps={{ options: { trailing: true } }} />
  );
}

TrailingTrue.storyName = 'Calls the callback after the wait time if trailing is true';
TrailingTrue.play = lokiPlayDecorator(
  async ({ canvasElement }: { canvasElement: HTMLElement }) => {
    const canvas = within(canvasElement);

    await userEvent.click(canvas.getByTestId('clickable'));
    await userEvent.click(canvas.getByTestId('clickable'));

    await waitFor(() => expect(canvas.getByTestId('count')).toHaveTextContent('Callback called 0 times'), { timeout: 200 });
    await waitFor(() => expect(canvas.getByTestId('count')).toHaveTextContent('Callback called 1 times'));
  }
);

export function LeadingAndTrailingTrue() {
  return (
    <ClickableComponent useDebounceCallbackProps={{ options: { leading: true, trailing: true } }} />
  );
}

LeadingAndTrailingTrue.storyName = 'Calls the callback immediately and after the wait time if both leading and trailing are true';
LeadingAndTrailingTrue.play = lokiPlayDecorator(
  async ({ canvasElement }: { canvasElement: HTMLElement }) => {
    const canvas = within(canvasElement);

    await userEvent.click(canvas.getByTestId('clickable'));
    await userEvent.click(canvas.getByTestId('clickable'));

    await waitFor(() => expect(canvas.getByTestId('count')).toHaveTextContent('Callback called 1 times'), { timeout: 200 });
    await waitFor(() => expect(canvas.getByTestId('count')).toHaveTextContent('Callback called 2 times'));
  }
);

export function LeadingAndTrailingFalse() {
  return (
    <ClickableComponent useDebounceCallbackProps={{ options: { leading: false, trailing: false } }} />
  );
}

LeadingAndTrailingFalse.storyName = 'Does not call the callback if both leading and trailing are false';
LeadingAndTrailingFalse.play = lokiPlayDecorator(
  async ({ canvasElement }: { canvasElement: HTMLElement }) => {
    const canvas = within(canvasElement);

    await userEvent.click(canvas.getByTestId('clickable'));
    await userEvent.click(canvas.getByTestId('clickable'));

    await waitFor(() => expect(canvas.getByTestId('count')).toHaveTextContent('Callback called 0 times'), { timeout: 200 });
    await waitFor(() => expect(canvas.getByTestId('count')).toHaveTextContent('Callback called 0 times'));
  }
);

export function DisableDebounce() {
  return (
    <ClickableComponent useDebounceCallbackProps={{ disableDebounce: true }} />
  );
}

DisableDebounce.storyName = 'Calls the callback immediately if disableDebounce is true';
DisableDebounce.play = lokiPlayDecorator(
  async ({ canvasElement }: { canvasElement: HTMLElement }) => {
    const canvas = within(canvasElement);

    await userEvent.click(canvas.getByTestId('clickable'));

    await waitFor(() => expect(canvas.getByTestId('count')).toHaveTextContent('Callback called 1 times'), { timeout: 0 });
  }
);
