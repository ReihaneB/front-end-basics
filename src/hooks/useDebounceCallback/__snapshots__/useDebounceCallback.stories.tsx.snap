// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hooks/useDebounceCallback CallAfterWaitTime play-test 1`] = `
<main class="inter-normal">
  <div>
    <button type="button"
            data-testid="clickable"
    >
      Click me
    </button>
    <p data-testid="count">
      Callback called 1 times
    </p>
  </div>
</main>
`;

exports[`Hooks/useDebounceCallback ClickMulipleTimes play-test 1`] = `
<main class="inter-normal">
  <div>
    <button type="button"
            data-testid="clickable"
    >
      Click me
    </button>
    <p data-testid="count">
      Callback called 1 times
    </p>
  </div>
</main>
`;

exports[`Hooks/useDebounceCallback DisableDebounce play-test 1`] = `
<main class="inter-normal">
  <div>
    <button type="button"
            data-testid="clickable"
    >
      Click me
    </button>
    <p data-testid="count">
      Callback called 1 times
    </p>
  </div>
</main>
`;

exports[`Hooks/useDebounceCallback DoNotCallBeforeWaitTime play-test 1`] = `
<main class="inter-normal">
  <div>
    <button type="button"
            data-testid="clickable"
    >
      Click me
    </button>
    <p data-testid="count">
      Callback called 1 times
    </p>
  </div>
</main>
`;

exports[`Hooks/useDebounceCallback LeadingAndTrailingFalse play-test 1`] = `
<main class="inter-normal">
  <div>
    <button type="button"
            data-testid="clickable"
    >
      Click me
    </button>
    <p data-testid="count">
      Callback called 0 times
    </p>
  </div>
</main>
`;

exports[`Hooks/useDebounceCallback LeadingAndTrailingTrue play-test 1`] = `
<main class="inter-normal">
  <div>
    <button type="button"
            data-testid="clickable"
    >
      Click me
    </button>
    <p data-testid="count">
      Callback called 2 times
    </p>
  </div>
</main>
`;

exports[`Hooks/useDebounceCallback LeadingTrue play-test 1`] = `
<main class="inter-normal">
  <div>
    <button type="button"
            data-testid="clickable"
    >
      Click me
    </button>
    <p data-testid="count">
      Callback called 1 times
    </p>
  </div>
</main>
`;

exports[`Hooks/useDebounceCallback Playground smoke-test 1`] = `
<main class="inter-normal">
  <div>
    <button type="button"
            data-testid="clickable"
    >
      Click me
    </button>
    <p data-testid="count">
      Callback called 0 times
    </p>
  </div>
</main>
`;

exports[`Hooks/useDebounceCallback TrailingTrue play-test 1`] = `
<main class="inter-normal">
  <div>
    <button type="button"
            data-testid="clickable"
    >
      Click me
    </button>
    <p data-testid="count">
      Callback called 1 times
    </p>
  </div>
</main>
`;
